### DISCUSSION TITLE
How can I updated the `Loaded` type of an entity after using `matching` or `loadItems`?

### DISCUSSION THREAD

**Discussion by SpencerKaiser** [mikro-orm/mikro-orm#4091] [Q&A] [answered]
I have an entity with a nested collection of items and I have a helper method to help me make sure relevant data is ALWAYS loaded before it goes too the front end (multiple endpoints return the same data + structure) but when I use `matching` or `loadItems`, the type sticks around as the original before the call... am I doing something wrong?

If that's expected behavior, how can I go about manually updating the type?

``` typescript
// author is of type `Loaded<Author, never>`

await const author.books.loadItems();
// or await const author.books.matching({genre: 'Fiction'});

// author is still of type `Loaded<Author, never>` but expected `Loaded<Author, 'books'>`
```

**Comment by B4nan** [comment, marked as answer]
What you can do now is defining new variable, the return type of those methods should be correct, or you can always use `em.populate`.

IIRC there is some TS feature that allows to augment types of existing properties, which could be a nice addition to v6, will take a look - just cracked the strict partial loading support so I am in a TS mood :]

edit: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions

**Reply by SpencerKaiser** [reply]
So it looks like the return type of `matching`, for example, is the `Collection` of loaded items... I tried to reassign that but I was skittish because if I use `author.books.set(await author.books.matching({...}));` and then accidentally persist/flush, won't that delete the ones that weren't matched?

Also somewhat related, but if I already have a populated collection, can I not use `matching` to filter? If I have `Loaded<Author, 'books'>` and I use `.matching` to with just `genre: 'Fiction'`, it doesn't look like anything changes ☹️ Maybe I'm misinterpreting how that should be used and it'll only do anything if the data is _**not**_ already loaded?

Also...
> so I am in a TS mood :]

Solid Friday Vibes 🚀

**Edit: the redundant `.matching` call _**does**_ make a query, it just doesn't remove the unrelated ones from the collection...**

**Reply by SpencerKaiser** [reply]
Also while you're looking for v6 additions along that same vein, it would be incredible if the `eager: true` option would propagate to the type of the entity, so if you created the entity with that setting for `book` the type would always be `Loaded<Author, 'books'>` 😬

**Reply by B4nan** [reply]
> Also while you're looking for v6 additions along that same vein, it would be incredible if the eager: true option would propagate to the type of the entity, so if you created the entity with that setting for book the type would always be Loaded<Author, 'books'> 😬

Hmm, maybe we could support this in the same way as `OptionalProps` or the PK type symbols, that could be actually quite simple.

> Also somewhat related, but if I already have a populated collection, can I not use matching to filter? If I have Loaded<Author, 'books'> and I use .matching to with just genre: 'Fiction', it doesn't look like anything changes ☹️ Maybe I'm misinterpreting how that should be used and it'll only do anything if the data is not already loaded?

Just do `coll.getItems().filter()` (or `loadItems` if you want to ensure it's loaded), the `matching` method has a different use case - it serves for partial loading of large collections, those you don't want to have in memory. The method by default does not store the state (it does not mutate the collection), you can use `store: true` in the options to do so.

**Reply by SpencerKaiser** [reply]
> Just do coll.getItems().filter() (or loadItems if you want to ensure it's loaded)

That's actually what I did initially which is how I arrived at the "How do I modify the type?" question 🤪 BUT that was an issue because I was trying to assign to `books` instead of using `books.set` 🤦‍♂️  

So I'm all set for now but I'd still love the features for changing the `Loaded` type after using something like `.populate`. Want me to open an issue for that?

**Reply by B4nan** [reply]
Looks like async type guards are unfortunately not supported (https://github.com/microsoft/TypeScript/issues/37681), but will see if I can make the `eager` work.

**Reply by B4nan** [reply]
<img width="508" alt="image" src="https://user-images.githubusercontent.com/615580/222834147-43fcfb07-18fe-4336-ae53-826cdc827bf3.png">

Yes, quite simple :]

**Reply by SpencerKaiser** [reply]
Ayyyy! Looks awesome 👌

**Reply by SpencerKaiser** [reply]
@B4nan I just tried this locally and the item that should have been a `LoadedReference` is still showing as a reference... it _does_ work for `ref: false`, but did you expect this to work for `ref: true` as well?

**Reply by SpencerKaiser** [reply]
And the flipside seems like maybe there's a bug there... without using `ref: true`, it seems relations appear to always be loaded 🤔 [Repro](https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading-xiwus3?file=basic.test.ts)

**Reply by B4nan** [reply]
You need to explicitly mark eager references with the `AddEager` symbol, don't expect decorator options to affect type-level information anyhow, that's simply not possible (unless something changed with ES decorators). Your test is either using extra `ref: true` on the author or you are missing `Ref` in the type - you either need both or none. With ts-morph the `ref: true` can be inferred from the type. And it does not fail because `author` is undefined, it fails because `author.books` is undefined - simply because it is not populated. I don't see any bug in there.

edit: sorry, not `AddEager`, the symbol is called `EagerProps`. it works like the `OptionalProps`. but I see your test has `eager: false` there, so now I am more confused what you were expecting from that

**Reply by SpencerKaiser** [reply]
Whoops, I missed the other `ref: true`... just added that back and you're correct on `author` not being `undefined`. Just [updated the repro](https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading-lst95p?file=entities%2Fbook.ts,basic.test.ts).

On that updated one, I expect `.books` to throw a TypeError because it's not loaded... do you think `Collection<Book, object>` is the right type for a property of something that is not eagerly loaded?

**Reply by SpencerKaiser** [reply]
Just tried adding `EagerProps` and didn't see any change on the types side...

```ts
[EagerProps]!: 'durationOption';

// ...

@ManyToOne({
    entity: () => DurationOption,
    ref: true,
    eager: true,
  })
  durationOption: Ref<DurationOption>;
```

Where I'm trying to use it doesn't have `$`, which is what I expected. I could be doing it wrong too, I'm not sure...

**Reply by B4nan** [reply]
Maybe this will help, the symbol should be optional - what you did results in runtime pollution of the prototype, while if you define it as optional it only affects the type.

```diff
-[EagerProps]!: 'durationOption';
+[EagerProps]?: 'durationOption';
```

**Reply by SpencerKaiser** [reply]
It still seems unhappy 😕 

<img width="787" alt="image" src="https://user-images.githubusercontent.com/6445731/230626755-d50c65bc-7e8a-4bf6-a188-c0f8c01a7659.png">

```ts
@ManyToOne({
  entity: () => DurationOption,
  ref: true,
  eager: true,
})
durationOption: Ref<DurationOption>;
```

**Reply by B4nan** [reply]
Make a repro for that I guess, hard to debug screenshots.

Btw I've invited you to the new getting started guide repo, feedback welcome on that. I'd like to have that ready for v6 release (and make it part of the docs, maybe the main part actually).

**Reply by SpencerKaiser** [reply]
Sounds good, I'll make one

And yea I saw that! I'll check it out later. If there aren't open issues to work from, anything specific you want help with?

**Reply by B4nan** [reply]
Maybe just go through what's there and if something is not clear enough, either PR it yourself or ping me somewhere. Same for the structure, or maybe I missed some important concept. I haven't touched it for some time, I am all in the v6 development now and want to focus on that now.

The plan is to have that in the docs, so the structure and how to split it into multiple pages is also something I am not yet sure about.

I could also image we replace some of the current docs sections with this guide.

**Reply by SpencerKaiser** [reply]
[Repro](https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading-6wrhdy?file=basic.test.ts)

Seems like it's only ever an issue when the eagerly loaded prop is on an entity that the entity you're using extends (See `CustomBaseEntity`)

**Reply by B4nan** [reply]
I don't see any bugs in there either, you only set `favoriteBook` as eager, but you expect `favoriteBook.author` to be loaded too? 

Moreover, you are trying to access `name` property on the reference, you either need the `load()` method or population with `$` property access - `Reference` class has no properties except the PK.

**Reply by B4nan** [reply]
https://stackblitz.com/edit/mikro-orm-v6-strict-partial-loading-xc2xv6?file=basic.test.ts

**Reply by SpencerKaiser** [reply]
No idea what I was trying to do there, but you're right... there's still an issue with `EagerProps` conflicting with the abstract entity, but I feel like that's an edge case on an edge case.

Based on our other discussion in #4213, I'm just going to stop using `eager` in general, so this is no longer an issue 😅
